onnxruntime/core/optimizer/matmul_scale_fusion.cc

```diff
@@ -0,0 +1,282 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/optimizer/matmul_scale_fusion.h"
+
+#include "onnx/defs/attr_proto_util.h"
+
+#include "core/common/optional.h"
+#include "core/framework/data_types_internal.h"
+#include "core/framework/tensorprotoutils.h"
+#include "core/graph/graph_utils.h"
+#include "core/graph/graph_viewer.h"
+#include "core/optimizer/utils.h"
+
+namespace onnxruntime {
+
+namespace {
+template <typename T>
+struct ExtractScalarAsFloatDispatchTarget {
+  Status operator()(const ONNX_NAMESPACE::TensorProto& tensor_proto, float& scalar_float) {
+    T scalar;
+    ORT_RETURN_IF_ERROR(utils::UnpackTensor(tensor_proto, &scalar, 1));
+    scalar_float = static_cast<float>(scalar);
+    return Status::OK();
+  }
+};
+
+optional<float> GetScalarConstantInitializer(const Graph& graph, const NodeArg& node_arg) {
+  const auto* initializer = graph_utils::GetConstantInitializer(graph, node_arg.Name());
```

Comment:
>GetConstantInitializer [](start = 41, length = 22)

we cannot rely on this function, right ?

graph.GetInitializedTensor() ?

Reply:
i think this is applicable here. using the ORT inference definition of "constant", we don't want to apply this fusion to an overridable initializer.

---
In reply to: [464691247](https://github.com/microsoft/onnxruntime/pull/4669#discussion_r464691247) [](ancestors = 464691247)

Reply:
ok


---
In reply to: [464692241](https://github.com/microsoft/onnxruntime/pull/4669#discussion_r464692241) [](ancestors = 464692241,464691247)